### PR TITLE
feat: reimplement ssldotcom-windows-sign

### PR DESCRIPTION
**Pull request by Gankra**

### CODE REVIEW

**Comment by Gankra** [comment]
Of note: the current implementation actually fails to pick up .ps1 scripts, which honestly I'm not sure matters but makes me a bit sad. This is a consequence of me requiring the feature only run on windows (could be changed) and .ps1 getting made on the global job, which is linux (could be changed?).

The main thing we care about is binaries and msi's.

File: cargo-dist/src/sign/ssldotcom.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        }
+    };
+    LocalAsset::write_new_all(
+        config.trim(),
+        tool_dir.join("conf/code_sign_tool.properties"),
+    )?;
+    Ok(())
+}
+
+// TODO: upstream fixes to axoasset to make RemoteAsset::copy work with .zips
```
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply cargo-dist/src/sign/ssldotcom.rs:192]: filed https://github.com/axodotdev/axoasset/issues/123

File: cargo-dist/src/sign/ssldotcom.rs [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        let credential_id = require_env_var("SSLDOTCOM_CREDENTIAL_ID")?;
+        let totp_secret = require_env_var("SSLDOTCOM_TOTP_SECRET")?;
+
+        // TODO: can we identify CI vs not to do this more carefully?
+        if username.is_empty()
+            || password.is_empty()
+            || credential_id.is_empty()
+            || totp_secret.is_empty()
+        {
+            warn!("skipping codesigning, some SSLDOTCOM env-vars aren't set");
```
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply cargo-dist/src/sign/ssldotcom.rs]: my latest commit should address this! good call

File: cargo-dist/src/sign/ssldotcom.rs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    const WINDOWS_URL: &str = "https://www.ssl.com/download/codesigntool-for-windows/";
+    const EXTRA_FETCH_DIR: &str = "_extra_tools";
+    const CODESIGNTOOL_SUBDIR: &str = "CodeSignTool";
+    const ZIP_NAME: &str = "CodeSignTool.zip";
+
+    let fetch_dir = dist_dir.join(EXTRA_FETCH_DIR);
+    let zip_path = fetch_dir.join(ZIP_NAME);
+    let unzipped_dir = fetch_dir.join(CODESIGNTOOL_SUBDIR);
+    let cmd = unzipped_dir.join(WINDOWS_CMD_NAME);
+    // let unzipped_dir = Utf8PathBuf::from("C:/Users/ninte/Downloads/CodeSignTool-v1.3.0-windows/");
```
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply cargo-dist/src/sign/ssldotcom.rs:164]: i'll remove it

**Review by mistydemeo — APPROVED** [review]
We can leave some of the followups, like the axoasset stuff, to followup PRs if we want to merge this now.

File: cargo-dist/src/sign/mod.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        let ssldotcom =
+            ssldotcom::CodeSignTool::new(host_target, dist_dir, ssldotcom_windows_sign)?;
+        Ok(Self { ssldotcom })
+    }
+
+    /// Sign a file
+    pub fn sign(&self, file: &Utf8Path) -> DistResult<()> {
+        if let Some(signer) = &self.ssldotcom {
+            let extension = file.extension().unwrap_or_default();
+            if let "exe" | "msi" | "ps1" = extension {
```
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply cargo-dist/src/sign/mod.rs:31]: this is a neat idea, if i can do it quick lets do it now

[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply cargo-dist/src/sign/mod.rs:31]: actually just filed https://github.com/axodotdev/cargo-dist/issues/1075

**Pull request merged**